src/sensor_hx71x.c

```diff
@@ -0,0 +1,267 @@
+    gpio_out_write(hx71x->sclk, 0);
+    hx71x_delay_no_irq(hx71x_get_time(), MIN_PULSE_TIME);
+    irq_enable();
+}
+
+// hx71x ADC query
+void
+hx71x_read_adc(struct hx71x_adc *hx71x, uint8_t oid)
+{
+    if (!hx71x_is_data_ready(hx71x)) {
+        hx71x_reschedule_timer(hx71x);
+        return;
+    }
+
+    // data is ready
+    int32_t counts = 0;
+    hx71x_time_t start_time = timer_read_time();
+    for (uint_fast8_t sample_idx = 0; sample_idx < 24; sample_idx++) {
+        hx71x_pulse_clocks(hx71x);
+        // read 2's compliment int bits
+        counts = (counts << 1) | gpio_in_read(hx71x->dout);
+    }
+
+    // bit bang 1 to 4 more bits to configure gain & channel for the next sample
+    for (uint8_t gain_idx = 0; gain_idx < hx71x->gain_channel; gain_idx++) {
+        hx71x_pulse_clocks(hx71x);
+    }
+
+    if (hx71x_is_data_ready(hx71x)) {
+        shutdown("HX71x Sensor reporting data ready after read");
+    }
+
+    hx71x_time_t time_diff = timer_read_time() - start_time;
+    if (time_diff >= MAX_READ_TIME) {
+        shutdown("HX71x Read took too long");
+    }
+
+    // extend 2's complement 24 bits to 32bits
+    if (counts >= 0x800000) {
+        counts |= 0xFF000000;
+    }
+
+    if (counts < -0x7FFFFF || counts > 0x7FFFFF) {
+        shutdown("HX71x Counts value out of range");
+    }
+    add_sample(hx71x, counts);
+
+    flush_samples(hx71x, oid);
+    hx71x_reschedule_timer(hx71x);
+}
+
+// Create a hx71x sensor
+void
+command_config_hx71x(uint32_t *args)
+{
+    struct hx71x_adc *hx71x = oid_alloc(args[0]
+                , command_config_hx71x, sizeof(*hx71x));
+    hx71x->timer.func = hx71x_event;
+    hx71x->pending_flag = 0;
+    uint8_t gain_channel = args[1];
+    if (gain_channel < 1 || gain_channel > 4) {
+        shutdown("HX71x gain/channel out of range 1-4");
+    }
+    hx71x->gain_channel = gain_channel;
+    hx71x->dout = gpio_in_setup(args[2], -1);
+    hx71x->sclk = gpio_out_setup(args[3], 0);
+}
+DECL_COMMAND(command_config_hx71x, "config_hx71x oid=%c gain_channel=%c"
+    " dout_pin=%u sclk_pin=%u");
+
+// start/stop capturing ADC data
+void
+command_query_hx71x(uint32_t *args)
+{
+    uint8_t oid = args[0];
+    struct hx71x_adc *hx71x = oid_lookup(oid, command_config_hx71x);
+    sched_del_timer(&hx71x->timer);
+    hx71x->pending_flag = 0;
+    hx71x->rest_ticks = args[1];
+    if (!hx71x->rest_ticks) {
+        // End measurements
+        return;
+    }
+    // Start new measurements
+    sensor_bulk_reset(&hx71x->sb);
+    hx71x_reschedule_timer(hx71x);
+}
+DECL_COMMAND(command_query_hx71x,
+             "query_hx71x oid=%c rest_ticks=%u");
+
+void
+command_query_hx71x_status(const uint32_t *args)
+{
+    uint8_t oid = args[0];
+    struct hx71x_adc *hx71x = oid_lookup(oid, command_config_hx71x);
+    const uint32_t start_t = timer_read_time();
+    uint8_t pending_bytes = 0;
+    pending_bytes = hx71x_is_data_ready(hx71x) * BYTES_PER_SAMPLE;
+    const uint32_t end_t = timer_read_time();
```

Comment:
It seems odd to perform the multiply in the timed critical section above.  A multiply can be several cycles on some chips.  (Realistically, gcc could move the code around, but no reason to hint to gcc that it should add code into the critical section.)

Reply:
I can make this a ternary

Reply:
Just to be clear, I was referring to performing operations between start_t and end_t.  I wasn't commenting on the operations themselves.

On second look though, you might want to do `irq_disable(); timer_read_time(); hx71x_is_data_ready(); irq_enable();` instead as that provides better timing and has less overhead than two `timer_read_time()` calls.  It's valid on the hx71x because `gpio_in_read()` is safe to run with irqs disabled.  When doing that, it is okay to pass `0` to the time diff field of `sensor_bulk_status()`.

Reply:
ok, sounds good